pydarn/io/borealis.py

```diff
@@ -0,0 +1,1135 @@
+            Dmap record
+        record_name : int
+            Record name for better error message information
+
+        Raises
+        -------
+        BorealisExtraField
+
+        """
+        file_struct = BorealisUtilities.dict_list2set(file_struct_list)
+        extra_fields = BorealisUtilities.dict_key_diff(record, file_struct)
+
+        if len(extra_fields) > 0:
+            raise borealis_exceptions.BorealisExtraFieldError(record_name,
+                                                              extra_fields)
+
+    @staticmethod
+    def incorrect_types_check(attributes_type_dict: dict,
+                              datasets_type_dict: dict,
+                              record: dict,
+                              record_name: int):
+        """
+        Checks if the file structure fields data type formats are correct
+        in the record.
+
+        Parameters
+        ----------
+        attributes_type_dict : dict
+            Dictionary with the required types for the attributes in the file.
+        datasets_type_dict : dict
+            Dictionary with the require dtypes for the numpy
+            arrays in the file.
+        record : dict
+            Dmap record
+        record_name : int
+            Record name for a better error message information
+
+        Raises
+        ------
+        BorealisFileFormatError
+        """
+        # check the attributes
+        incorrect_types_check = {param: str(attributes_type_dict[param])
+                                 for param in attributes_type_dict.keys()
+                                 if type(record[param]) !=
+                                 attributes_type_dict[param]}
+        # the rest of the parameters in the
+        # record are numpy arrays (HDF5 datasets)
+        incorrect_types_check.update({param: 'np.ndarray of ' +
+                                      str(datasets_type_dict[param])
+                                      for param in datasets_type_dict.keys()
+                                      if record[param].dtype.type !=
+                                      datasets_type_dict[param]})
+        if len(incorrect_types_check) > 0:
+            raise borealis_exceptions.BorealisDataFormatTypeError(incorrect_types_check,
+                                                                  record_name)
+
+
+class BorealisRead():
+
+    def __init__(self, borealis_file: str):
+        """
+        Reads SuperDARN Borealis file types into a
+        dictionary for reading methods.
+
+        Parameters
+        ----------
+        borealis_file : str
+            file name containing Borealis hdf5 data.
+
+        Raises
+        ------
+        OSError
+            Unable to open file
+        """
+        self.borealis_file = borealis_file
+
+        with h5py.File(self.borealis_file, 'r') as f:
+            self._group_names = sorted(list(f.keys()))
+            # list of group keys in the HDF5 file, to allow partial read.
```

Comment:
group names are provided as a property so that if desired you could potentially do a partial read by providing the record name. You can't do a read of a single record unless you know the group name (if using this deepdish library). I didn't fully set that up here (reading the record is a private function), but I still think the group names are a useful property of the class